### PR TITLE
[ParseableInterfaces] Refine cache keys and dependency checks

### PR DESCRIPTION
**Pull request by harlanhaskins**
Hashing the contents of the interface files for dependency checking is overkill. In practice, size and last modification time are enough to determine if a file has changed on disk, and therefore should be rebuilt.

As for cache keys, previously, we included the PCH hash components in the cache key. While they didn’t do any harm, they didn’t contribute any unique information about the module in question.

Additionally, passing the effective language version in means that each dependency that uses a different -swift-version would re-compile all of its dependencies. This is unfortunate, as that means the standard library is recompiled potentially several times.

rdar://46503065

### CODE REVIEW

File: test/ParseableInterface/ModuleCache/module-cache-intermediate-mtime-change-rebuilds-only-intermediate.swift [hunk 1]
```diff
@@ -27,12 +27,11 @@
 // RUN: %{python} %S/Inputs/make-old.py %t/modulecache/OtherModule-*.swiftmodule
 //
 //
-// Actual test: Change a byte in OtherModule.swiftinterface, check we only rebuild its cached module.
+// Actual test: Change the mtime of OtherModule.swiftinterface, check we only rebuild its cached module.
```
Comment (jrose-apple) [inline test/ParseableInterface/ModuleCache/module-cache-intermediate-mtime-change-rebuilds-only-intermediate.swift]: Are there already tests for changing the size but keeping the mtime constant?

File: lib/Frontend/ParseableInterfaceSupport.cpp [hunk 2]
```diff
@@ -124,20 +125,23 @@ static std::string getCacheHash(ASTContext &Ctx,
                                 const CompilerInvocation &SubInvocation,
                                 StringRef InPath) {
   // Start with the compiler version (which will be either tag names or revs).
-  std::string vers = swift::version::getSwiftFullVersion(
-      Ctx.LangOpts.EffectiveLanguageVersion);
-  llvm::hash_code H = llvm::hash_value(vers);
+  // Explicitly don't pass in the "effective" language version -- this would
+  // mean modules built in different -swift-version modes would rebuild their
+  // dependencies.
+  llvm::hash_code H = hash_value(swift::version::getSwiftFullVersion());
```
Comment (jrose-apple) [inline lib/Frontend/ParseableInterfaceSupport.cpp]: Can you add a test that this doesn't happen anymore?

File: test/ParseableInterface/ModuleCache/module-cache-intermediate-mtime-change-rebuilds-only-intermediate.swift (hunk 1 above)
Comment (harlanhaskins) [reply test/ParseableInterface/ModuleCache/module-cache-intermediate-mtime-change-rebuilds-only-intermediate.swift]: Yep, for both the leaf and intermediate modules.

File: lib/Frontend/ParseableInterfaceSupport.cpp (hunk 2 above)
Comment (harlanhaskins) [reply lib/Frontend/ParseableInterfaceSupport.cpp]: Ah! Yes, thanks for the reminder.

Comment (harlanhaskins) [reply lib/Frontend/ParseableInterfaceSupport.cpp]: Added the test. Fails with an older build, succeeds with this one.

**Review by jrose-apple — APPROVED** [review]

**Pull request merged**